frontend_tests/node_tests/stream_edit.js

```diff
@@ -0,0 +1,271 @@
+            assert(sort_streams_called);
+        })();
+
+        (function test_updater() {
+            function number_of_pills() {
+                const pills = stream_edit.pill_widget.items();
+                return pills.length;
+            }
+
+            assert.equal(number_of_pills(), 0);
+            config.updater.call(fake_this, denmark);
+            assert.equal(number_of_pills(), 1);
+            fake_this.query = me.email;
+            config.updater.call(fake_this, me);
+            assert.equal(number_of_pills(), 2);
+            fake_this.query = "#Denmark";
+        })();
+
+        (function test_source() {
+            const result = config.source.call(fake_this);
+            const taken_ids = stream_pill.get_stream_ids(stream_edit.pill_widget);
+            const stream_ids = result.map((stream) => stream.stream_id).sort();
+            let expected_ids = subs.map((stream) => stream.stream_id).sort();
+            expected_ids = expected_ids.filter((id) => !taken_ids.includes(id));
+            assert.deepEqual(stream_ids, expected_ids);
+        })();
+
+        input_typeahead_called = true;
+    };
+
+    function get_users_for_subscriber_typeahead() {
+        const sub = subs.get_sub_by_id(netherlands.stream_id);
+        const potential_subscribers = stream_data.potential_subscribers(sub);
+        return user_pill.filter_taken_users(potential_subscribers, stream_edit.pill_widget);
+    }
+
+    const opts = {source: get_users_for_subscriber_typeahead, stream: true};
+    pill_typeahead.set_up(input_field_stub, stream_edit.pill_widget, opts);
+    assert(input_typeahead_called);
+
+    let add_subscribers_handler = $(subscriptions_table_selector).get_on_handler(
+        "submit",
+        ".subscriber_list_add form",
+    );
+    const fake_this = $.create(".subscription_settings");
+    fake_this.closest = function () {
+        return 1;
+    };
+    const event = {
+        target: fake_this,
+        preventDefault: () => {},
+    };
+
+    called = false;
+    add_subscribers_handler(event);
+    assert(called);
+
+    add_subscribers_handler = $(subscriptions_table_selector).get_on_handler(
+        "keyup",
+        ".subscriber_list_add form",
+    );
+    event.which = 13;
+
+    called = false;
+    user_pill.get_user_ids = () => [];
+    add_subscribers_handler(event);
+    assert(called);
+
+    // No request is sent when there are no users to subscribe.
+    called = false;
+    stream_pill.get_user_ids = () => [];
+    add_subscribers_handler(event);
+    assert(!called);
+
+    // No request is sent if we try to subscribe ourselves
+    // only and are already subscribed to the stream.
+    called = false;
+    user_pill.get_user_ids = () => [me.user_id];
+    add_subscribers_handler(event);
+    assert(!called);
+
+    called = false;
+    user_pill.get_user_ids = () => [mark.user_id, me.user_id];
+    add_subscribers_handler(event);
+    assert(called);
+});
```

Comment:
I feel like this test is too mock-heavy.  Ideally, we should be doing end-to-end tests of the main logic, e.g. "Type these 5 characters, verify the typeahead options are X, Y, Z in that order".  Then "Select Y, verify it creates a pill with Z data".  Reading this test it seems like all we're doing is getting test "coverage" -- i.e. checking all the lines execute -- but not verifying much about the actual output.

(I could be missing something as I'm mostly skimming these tests)

https://zulip.readthedocs.io/en/latest/testing/philosophy.html

@showell or @aero31aero may be able to provide more advice on setting this test up properly; our main typeahead tests balance this stuff well.